sdk/examples/xo_rust/src/handler/handler.rs

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2018 Bitwise IO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+use sawtooth_sdk::messages::processor::TpProcessRequest;
+use sawtooth_sdk::processor::handler::ApplyError;
+use sawtooth_sdk::processor::handler::TransactionContext;
+use sawtooth_sdk::processor::handler::TransactionHandler;
+
+use handler::helpers;
+use handler::payload::XoPayload;
+use handler::state::Game;
+use handler::state::XoState;
+use handler::state::get_xo_prefix;
+
+pub struct XoTransactionHandler {
+    family_name: String,
+    family_versions: Vec<String>,
+    namespaces: Vec<String>,
+}
+
+impl XoTransactionHandler {
+    pub fn new() -> XoTransactionHandler {
+        XoTransactionHandler {
+            family_name: String::from("xo"),
+            family_versions: vec![String::from("1.0")],
+            namespaces: vec![String::from(get_xo_prefix().to_string())],
+        }
+    }
+}
+
+impl TransactionHandler for XoTransactionHandler {
+    fn family_name(&self) -> String {
+        self.family_name.clone()
+    }
+
+    fn family_versions(&self) -> Vec<String> {
+        self.family_versions.clone()
+    }
+
+    fn namespaces(&self) -> Vec<String> {
+        self.namespaces.clone()
+    }
+
+    fn apply(
+        &self,
+        request: &TpProcessRequest,
+        context: &mut TransactionContext,
+    ) -> Result<(), ApplyError> {
+        let header = &request.header;
+        let signer = match &header.as_ref() {
+            Some(s) => &s.signer_public_key,
+            None => return Err(helpers::app_err("Invalid header")),
+        };
+
+        let payload = match XoPayload::new(&request.payload) {
+            Ok(opt) => match opt {
+                Some(pay) => pay,
+                None => return Err(helpers::app_err("Bad payload")),
+            },
+            Err(_) => return Err(helpers::app_err("Bad payload")),
```

Comment:
This throws the cause (the error causing the new error) away; it would be better to propagate the original error up or work the cause's error message into the new error message.